### DISCUSSION TITLE
Migrate `InboundService` `Mempool`, `SyncService` to `zebra-?` crate

### DISCUSSION THREAD

**Discussion by dconnolly** [ZcashFoundation/zebra#3598] [Q&A] [answered]
> zebrad and zebra-rpc inter-dependency (e.g., mempool service's Request type)

What should we call it? Any weird things to keep in mind?

**Comment by dconnolly** [comment]
Idea: `zebra-services`

**Reply by dconnolly** [reply]
Also, `zebrad-services`?

**Reply by teor2345** [reply]
Seems very generic, we also have network and state services.

**Reply by jvff** [reply]
I agree it's rather generic, but we are including services which don't share a common description :/

If it helps, my thought process is that I'd rather have a generic broad name than a specific but confusing or fragile name.

**Reply by teor2345** [reply]
What about `zebra-node-services`?

(I don't have a strong opinion here.)

**Comment by teor2345** [comment]
Ideas:
- `zebra-sync`
- `zebra-share`
- `zebra-distribute` (not quite right, too long?)

**Comment by teor2345** [comment]
I'd also like to move the `update_progress` function to the new crate. It's useful functionality that doesn't depend on anything else in `zebrad`.

https://github.com/ZcashFoundation/zebra/blob/1896943f62fdb0240a0b4db55fa8c10a7058c79b/zebrad/src/commands/start.rs#L303-L308

If we move all these components, then `zebrad` is left with setup, the main loop, argument processing, and monitoring. Also, anyone who wants to create a different app can do it using the new crate.

**Comment by jvff** [comment]
We could also create more than one crate, like `zebra-mempool` and `zebra-sync` 🤔

**Reply by teor2345** [reply]
We could delay that decision, and create more crates later as we need them.

**Comment by jvff** [comment]
Another idea, based on @dconnolly's suggestion to use common traits yesterday at the meeting: have a `zebra-services` crate that **only** includes `Request` and `Response` types for **all** services (network, state, mempool, inbound, sync, ...), and maybe it could be a crate that depends on little else (perhaps only `zebra-chain`?).

**Reply by teor2345** [reply]
Splitting out all the Zebra the `Request` and `Response` types could be a really complicated change. It also expands the scope - we only need to move the mempool and its dependencies for the lightwalletd work.

We could delay this decision, and do the `Request` and `Response` split when we need to.

**Reply by jvff** [reply]
It's true that it expands the scope, but we could limit it by just moving the mempool `Request` and `Response` for now, so it would be an even smaller change.

I might be wrong, but my intuition is that it would also be simpler longer term :thinking:

**Reply by jvff** [reply]
Thinking more about this, I think it's the simplest solution that unblocks me while still keeping our future options open. I can:

1. Create a `zebra-services` crate
2. Move only the mempool `Request` and `Response` to it
3. Update `zebrad` to import those types from the new crate

In the future we can still:

1. Rename the crate
2. Add more `Request` and `Response` types as I suggested
3. Move the actual service implementations to the new crate

**Reply by teor2345** [reply]
> we could limit it by just moving the mempool `Request` and `Response` for now

We also need to move `Gossip`, and replace the `MempoolError` in `Response::Queued` with `BoxError`.
(So we don't pull in all the mempool storage error types.)

I agree this is the minimal change we need.

**Reply by teor2345** [reply]
How do you feel about the name `zebra-node-services`?

This distinguishes the new crate from the low-level network, state, and consensus crates. 

(I don't have a strong opinion here.)

**Reply by jvff** [reply]
No strong opinion either 😅 

Maybe it's best to call it that now to reduce confusion with larger states, and we can change it in the future if we think of a better name 🤔

**Reply by dconnolly** [reply]
sgtm 👍

**Comment by teor2345** [comment]
These crate changes are blocking the mempool lightwalletd APIs, which are otherwise ready to go.

I was hoping we could quickly:
- pick a name
- create a new crate
- move a few files
- update dependency paths
and then get on with the actual lightwalletd work.

@conradoplg how did you want to pick a name here?

**Reply by teor2345** [reply]
I think we might have reached a conclusion in this thread:
https://github.com/ZcashFoundation/zebra/discussions/3598#discussioncomment-2228922

**Comment by gustavovalverde** [comment, marked as answer]
If there's consensus and a PR, I'd recommend turning this discussion into a ticket (there's an action here for that), so it can be added to ZenHub for tracking and closed when the PR is merged @jvff @teor2345

**Reply by dconnolly** [reply]
I think `zebra-node-services` works: https://github.com/ZcashFoundation/zebra/discussions/3598#discussioncomment-2232776
